File: 2022/java/src/main/Day6.java
package main;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.HashSet;

public class Day6 {
  public static void main(String[] args) throws Exception {
    var line = Files.readAllLines(Path.of("src/resources/day6.txt")).get(0);
    System.out.println(getMarker(line, 4));
    System.out.println(getMarker(line, 14));
  }

  private static int getMarker(String line, int distinctNumbers) {
    for (int i = 0; i < line.length() - distinctNumbers - 1; i++) {
      var set = new HashSet<>(Arrays.asList(line.substring(i, i + distinctNumbers).split("")));
      if (set.size() == distinctNumbers)
        return i + distinctNumbers;
    }
    return 0;
  }
}

File: 2022/kotlin/src/main/kotlin/Day6.kt
fun main() {
    val line = readInput("day6")[0]
    println(getMarker(line, 4))
    println(getMarker(line, 14))
}

private fun getMarker(line: String, distinctNumbers: Int): Int {
    for (i in 0..<line.length - distinctNumbers - 1) {
        val size = line.substring(i, i + distinctNumbers)
            .split("")
            .filter { it != "" }
            .toHashSet().size
        if (size == distinctNumbers) return i + distinctNumbers
    }
    return 0
}
